### DISCUSSION TITLE
using the same useQuery key in different views

### DISCUSSION THREAD

**Discussion by emehran** [TanStack/query#1813] [Q&A]
Hey,
for a breadcrumb situation, I need to run the same query key like ['name', 100] in different views of the app. Ideally, I like to benefit from the cache, and for the next views other than the first useQuery, first load it instantly from cache and then based on my config trigger a fetch.

based on what I'm experiencing using the same key, seems like every time I visit the next view, it fetches it again without first loading it from the cache.

my code is like

View1: UseQuery(['name', id]....
View2: UseQuery(['name', id]....

for the same id, it blinks and fetches all the time. 
Am I doing it with the right approach?

**Comment by TkDodo** [comment]
Looks like those views don’t share the same QueryClientProvider. Do you give each view their own, and thus their own cache?

**Reply by emehran** [reply]
the Provider wraps all components in App.tsx, so basically they share the same provider. So the answer is no. 

If I give the first useQuery (['a', id], fn) and the second (['b', id], fn), it works fine as it treats the same GET fetch as separate queries and creates cache for each one separately, but I want to use the same key for the same query

**Reply by TkDodo** [reply]
> using the same key, seems like every time I visit the next view, it fetches it again without first loading it from the cache.

I would need to see that in a codesandbox please. If you call `useQuery` with the same `key`, and the cache has data for that `key`, you _should_ get that data returned immediately. All of the examples show basically that behaviour...

**Reply by emehran** [reply]
I figured out why it was happening.  In the second view, the async data function was returning the same data with some modifications in the object properties like

first function: 
return {a, b}

second function: 
return {A:a, B:b}

so the cached data was different and it was behaving like that.

anyway it's resolved, thanks

**Reply by trongtai37** [reply]
Hi @TkDodo, I have some following up questions.
```ts
// file: keys.ts
export const queryKey = 'queryKey'

// file: ComponentA.tsx
useQuery(queryKey, functionA)

// file: ComponentB.tsx
useQuery(queryKey, functionB)
``` 

- Where was functionA and functionB stored, `QueryObserver` level or `QueryCache` level?
- When I want to `queryClient.invalidateQueries([queryKey])`, which function is used to fetch new data and why?

**Reply by bystrovo** [reply]
interested in the answer as well

**Reply by TkDodo** [reply]
> Where was functionA and functionB stored, QueryObserver level or QueryCache level?

cache

> When I want to queryClient.invalidateQueries([queryKey]), which function is used to fetch new data and why?


undefined behaviour, not recommended. My guess is the "later" one, but please don't do this. One query key needs to be paired with exactly one queryFn.

**Reply by trongtai37** [reply]
Thanks @TkDodo, now I know why this [eslint rule](https://tanstack.com/query/latest/docs/react/eslint/exhaustive-deps) is so important :).
@bystrovo FYI I made a codesandbox to test it too https://codesandbox.io/p/sandbox/zealous-mcnulty-jjfqyn

**Reply by hamzadarej** [reply]
i had the same problem, query refetch when i change the view using the same key: i fixed this using this config { refetchOnWindowFocus: false, enabled : !queryCache.getQueryData(['key', var ])
